Document how to migrate a standalone role to a collection (#68687)
ansible/ansible: docs/docsite/rst/dev_guide/migrating_roles.rst

.. _migrating_roles:

***************************************
Migrating Roles to Roles in Collections
***************************************

> Collections can contain roles, modules, plugins, and playbooks. See :ref:`developing_collections` for details.

Reviewer: I'd change the topic sentence to be more directly about migrating - something like, "If you maintain standalone roles on Galaxy, you can migrate them into collections." Plus a sentence about why you would want to do that . . .

You can migrate an existing standalone role into a collection-based role.

.. contents::
   :local:
   :depth: 2

.. _roles_in_collections:

Comparing standalone roles to collection roles
===============================================

:ref:`Standalone roles <playbooks_reuse_roles>` have the following directory structure:

.. code-block:: bash
  :emphasize-lines: 5,7

    role/
    ├── defaults
    ├── files
    ├── handlers
    ├── library

>     ├── meta

Reviewer 2: module_utils

- Collection roles cannot contain a :file:`library/` directory with custom modules. Move modules to the :file:`plugins/modules/` directory of the collection. You can reference these collection modules in any of the roles contained within a collection, or externally using the fully qualified collection name (:abbr:`FQCN (Fully Qualified Collection Name)`).
- Modules inside a collection need :abbr:`FQCN (Fully Qualified Collection Name)` paths for their Python imports. See :ref:`migrating_plugins_collection`.
- Standalone roles can optionally encapsulate any plugin in a :file:`PLUGINTYPE_plugins/` directory (such as :file:`filters_plugin/`). For roles in a collection, these plugins go into their respective :file:`plugins/PLUGINTYPE/` directory (such as :file:`plugins/filter/`) in an collection. You reference these plugins with their :abbr:`FQCN (Fully Qualified Collection Name)`.

.. note::

	 In standalone roles, some of the plugin directories referenced their plugin types in the plural sense; this is not the case in collections.

Migrating a role to a collection
================================

> 

Reviewer: Maybe add an overview of the steps, plus a sub-heading for `Create a collection`?

> In order to migrate from a standalone role to a collection role, create a collection with the ``ansible-galaxy collection`` CLI command. You need a `Galaxy namespace <https://galaxy.ansible.com/docs/contributing/namespaces.html>`_ to import this collection to Galaxy.

Reviewer: ```suggestion
To migrate from a standalone role to a collection role, create a collection with the ``ansible-galaxy collection`` CLI command. You need a `Galaxy namespace <https://galaxy.ansible.com/docs/contributing/namespaces.html>`_ to import this collection to Galaxy.
```

.. code-block:: bash

  $ ansible-galaxy collection init mynamespace.mycollection

This creates the collection directory structure. The :file:`mynamespace/mycollection/plugins/` directory contains a :file:`README.md` file that explains the various types of plugins that the collection can contain within optionally created subdirectories.

> Migrating a standalone role without plugins or modules

Reviewer: I think I would re-work this section as a procedure. The steps would be:
1. create a collection
2. migrate the role
3. if your role contains modules or plugins, move them
4. if you moved modules or plugins, update them (with subsections on custom module_utils and init.py)

------------------------------------------------------

If you have a standalone role that does not contain any custom modules or plugins, copy the entire standalone role directory into the :file:`roles/` subdirectory of the collection. You can then reference the role with the :abbr:`FQCN (Fully Qualified Collection Name)`.

.. code-block:: bash

  $ mkdir mynamespace/mycollection/roles/my_role/
  $ cp -r /path/to/standalone/role/mynamespace/my_role/\* mynamespace/mycollection/roles/my_role/

The following example uses this collection role in a playbook:

.. code-block:: yaml

  ---
  - name: example role by FQCN
    hosts: some_host_pattern
    tasks:
      - name: import FQCN role from a collection
        import_role:
          name: mynamespace.mycollection.my_role

Note in this example that the type of content from inside a collection is inferred contextually. See :ref:`collections_using_playbook` for more details.

.. _migrating_plugins_collection:

Migrating plugins and modules
-----------------------------

> Migrating plugins and modules from a standalone role to a collection requires a few more steps.

Reviewer: I'd put an example of actually moving the plugins and modules here - something like `mv mynamespace/mycollection/roles/my_role/library/whatever.whatever mynamespace/mycollection/plugins/whatever` (<<< is an incorrect or at least incomplete example, but you get the idea).

Custom module_utils
^^^^^^^^^^^^^^^^^^^

> If you have a module that uses a custom ``module_utils``, you could previously address it in the top level ``ansible.module_utils`` Python namespace. This is no longer the case. The top level Python namespace no longer merges the Ansible internal Python namespace with external content and merging external content from collections into the Ansible internal Python namespace is not supported. This is explained below but see :ref:`collection_module_utils` for more details.

Reviewer: ```suggestion
If any of your custom modules use a custom module utility, once you migrate to a collection you cannot address the module utility in the top level ``ansible.module_utils`` Python namespace. Ansible does not merge content from collections into the the Ansible internal Python namespace. You must update any Python import statements that refer to custom module utilities when you migrate your custom content to collections. See :ref:`collection_module_utils` for more details.
```

  $spec = @{
	  name = @{ required = $true; type = "str" }
  	state = @{ required = $true; choices = @("present", "absent") }
  }
  $module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)

  Invoke-HyperVFunction -Name $module.Params.name

  $module.ExitJson()

Importing from __init__.py
^^^^^^^^^^^^^^^^^^^^^^^^^^

Because of the way that the CPython interpreter does imports, combined with the way the Ansible plugin loader works, if your custom embedded module or plugin requires importing something from an :file:`__init__.py` file, that also becomes part of your collection. You can either originate the content inside a standalone role or use the file name in the Python import statement. The following example is an :file:`__init__.py` file that is part of a callback plugin found inside a collection named ``ansible_example.community``.

.. code-block:: python

  from ansible_collections.ansible_example.community.plugins.callback.__init__ import CustomBaseClass

Example: Migrating a standalone role with modules to a collection
=================================================================

> In this example we have a standalone role called ``my-standalone-role.webapp`` to emulate a standalone role that contains dashes in the name (which is not valid in collections). This standalone role contains a custom module in the ``library/`` directory called ``manage_webserver``.

Reviewer: The fact that role names cannot have dashes in collections is kind of buried here. Can we mention that earlier/in a more prominent way?

.. note::

	Roles in collections cannot have hyphens in the role name. Rename any such roles to use underscores instead.

.. contents::
   :local:
   :depth: 1

.. _roles_in_collections:

Comparing standalone roles to collection roles
===============================================

:ref:`Standalone roles <playbooks_reuse_roles>` have the following directory structure:

.. code-block:: bash
  :emphasize-lines: 5,7,8

    role/
    ├── defaults
    ├── files
    ├── handlers
    ├── library
    ├── meta
    ├── module_utils
    ├── [*_plugins]
    ├── tasks
    ├── templates
    ├── tests
    └── vars

The highlighted directories above will change when you migrate to a collection-based role. The collection directory structure includes a :file:`roles/` directory:

.. code-block:: bash

    namespace/
    └── name/
      ├── docs/
      ├── galaxy.yml
      ├── plugins/
      │   ├── modules/
      │   │   └── module1.py
      │   ├── inventory/
      │   └── .../
      ├── README.md
      ├── roles/
      │   ├── role1/
      │   ├── role2/
      │   └── .../
      ├── playbooks/
      │   ├── files/

>       │   ├── vars/

Reviewer 2: defaults

      │   ├── templates/
      │   └── tasks/
      └── tests/

A collection can contain one or more roles in the :file:`roles/` directory and these are almost identical to standalone roles with the following exceptions:

+----------------------------------------+------------------------------------------------------------------------------+
| standalone role                        | collection role                                                              |
+========================================+==============================================================================+
| :file:`library/` with custom modules   | - Move to :file:`plugins/modules/`                                           |

> |                                        | - Use :abbr:`FQCN (Fully Qualified Collection Name)` external to collection  |

Reviewer 3: This is the first mention of FQCN on this page, and already, even after reading the [developing collections](https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html) guide linked earlier, the statement "Use `FQCN (Fully Qualified Collection Name)` external to collection" doesn't mean a whole lot to me.

Does this mean when I move a module from `library/` to `plugins/modules`, it needs to be named on the filesystem using the FQCN? Or do I need to change things inside the plugin to use an FQCN?

Additionally, it's a little confusing to me—do I move a file inside a role into a collection _role_ `plugins/modules` directory, or do I have to move it in to a _collection_ `plugins/modules` directory? The way this table is formatted and with the text above, it looks like the module should still be located physically within the role directory, but it should be in the role directory's `plugins/modules` directory instead of the role directory's `library` directory.

Reviewer 3: Additionally "external to collection" doesn't make sense to me—how is the FQCN different whether I'm in the collection or external to it? Isn't it always "namespace.name"?

One other question I ponder is what about collections that are not meant for Galaxy/Automation Hub—do they still need to have some sort of made-up namespace and name? Or can I have a collection called "stuff" that is not namespaced? (Haven't tested this, but wondering—since standalone roles can exist without a namespace.)

Author: I took out the table as it wasn't proving a good summary of what needs to happen. I also described  FQCN upfront. Not sure how to handle what you pondered in the end there about non-galaxy collections....

+----------------------------------------+------------------------------------------------------------------------------+

> |``ansible.module_utils`` in Python      + Use :abbr:`FQCN (Fully Qualified Collection Name)` for Python imports        |

Reviewer 3: Also since we're talking about roles still (a standalone role vs a role in a collection), what other content within the role directory besides plugins (which I believe have to actually move out of the role and into the collection proper) would need to use `ansible.module_utils`?

If this is more of a guide for "how to move plugins out of roles into collection plugins directories", then maybe this section/table should be re-introduced as

"roles can have the same structure inside collections as when they're standalone, but you have to move all the plugins and supporting Python code out of the role and into the collection, renaming and modifying the Python code like so:"

Author: snipped the table entirely. I think this is covered in the steps but let me know if anything is still missing.

| Plugins in :file:`PLUGINTYPE_plugins/` + - Move to :file:`plugins/PLUGINTYPE/`  (such as :file:`plugins/filter/`)     |
|                                        | - Use :abbr:`FQCN (Fully Qualified Collection Name)` external to collection  |
+----------------------------------------+------------------------------------------------------------------------------+
| :file:`module_utils/`                  + - Move to :file:`plugins/module_utils/`                                      |
|                                        | - Use :abbr:`FQCN (Fully Qualified Collection Name)` external to collection  |
+----------------------------------------+------------------------------------------------------------------------------+

.. note::

	 In standalone roles, some of the plugin directories referenced their plugin types in the plural sense; this is not the case in collections.

Migrating a role to a collection
================================

To migrate from a standalone role to a collection role:

1. Create a collection. You need a `Galaxy namespace <https://galaxy.ansible.com/docs/contributing/namespaces.html>`_ to import this collection to Galaxy.

.. code-block:: bash

>   $ ansible-galaxy collection init mynamespace.mycollection

Reviewer 3: I ran into an issue here—since this command doesn't respect the `collections_paths` you have configured, it will dump a collection structure into `./mynamespace/mycollection`. Since that is not in a collections_paths with the magic folder `ansible_collections`, any playbooks you have locally that refer to the new collection won't be able to find it—unless you manually create an `ansible_collections` directory in a `collections_paths` location (which is not documented), then change directories to that location, then continue with this guide from that point onward.

Reviewer 2: we can distinguish 'setting up collection for dev' vs execution of said collection (which needs to be in known execution path with 'magic dirs' on top.

Author: not sure what to change in the doc here?

This creates the collection directory structure.

2. Copy the standalone role directory into the :file:`roles/` subdirectory of the collection.

.. code-block:: bash

  $ mkdir mynamespace/mycollection/roles/my_role/

>   $ cp -r /path/to/standalone/role/mynamespace/my_role/\* mynamespace/mycollection/roles/my_role/

Reviewer 2: note that this might require a rename of role since naming rules in collection differ from 'non collection roles'

4. Move any modules to the :file:`plugins/modules/` directory.

.. code-block:: bash

  $ mv -r mynamespace/mycollection/roles/my_role/library/\* mynamespace/mycollection/plugins/modules/

4. Move any other plugins to the appropriate :file:`plugins/PLUGINTYPE/` directory. The :file:`mynamespace/mycollection/plugins/README.md` file explains the types of plugins that the collection can contain within optionally created subdirectories.

.. code-block:: bash

  $ mv -r mynamespace/mycollection/roles/my_role/filter_plugins/\* mynamespace/mycollection/plugins/filter/

> 

Reviewer 2: make note here, 'moving the plugins' is normally not enough, more work is needed as detailed below.

5. Change any references to the role, modules, or plugins with the :abbr:`FQCN (Fully Qualified Collection Name)`.

.. code-block:: yaml

  ---
  - name: example role by FQCN
    hosts: some_host_pattern
    tasks:
      - name: import FQCN role from a collection
        import_role:
          name: mynamespace.mycollection.my_role

> Note in this example that the type of content from inside a collection is inferred contextually. See :ref:`collections_using_playbook` for more details.

Reviewer 3: > inferred contextually

Not sure what is meant here? Does it just mean "when you call a role it will call a role, and when you call a module it will call a module?"

Not sure if this is the best way to say that, more like "use FQCN to call the roles / plugins / modules in your collection" or something like that. Or say nothing at all?

Also, for those (like me) who don't want to also have to rewrite all their playbooks after doing this (or just like using the shorthand instead of getting carpal tunnel from writing FQCNs on every task), what about showing the `collections:` keyword example as mentioned here: https://docs.ansible.com/ansible/latest/user_guide/collections_using.html#using-collections-in-a-playbook

```yaml
---
- name: example role by FQCN
  hosts: some_host_pattern
  collections:
    - mynamespace.mycollection
  tasks:
    - name: import role from a collection
      import_role:
        name: my_role
```

.. note::

  This example changed the original source file ``manage_webserver.py`` to the destination file ``manage.py``. This is optional but the :abbr:`FQCN (Fully Qualified Collection Name)` provides the ``webserver`` context as ``acme.webserver.manage``.

4. Change ``manage_webserver`` to ``acme.webserver.manage`` in :file:`tasks/` files in the role ( for example, ``my-standalone-role.webapp/tasks/main.yml``) and any use of the original module name.

.. note::

  This name change is only required if you changed the original module name, but illustrates content referenced by :abbr:`FQCN (Fully Qualified Collection Name)` can offer context and in turn can make module and plugin names shorter. If you anticipate using these modules independent of the role, keep the original naming conventions. Users can add the  :ref:`collections keyword <collections_using_playbook>` in their playbooks. Typically roles are an abstraction layer and users won't use components of the role independently.

Supporting standalone roles and migrated collection roles in a downstream RPM
=============================================================================

> A standalone role can co-exist with its collection role counterpart (for example, as part of a support lifecycle of a product). This should only be done for a transition period, but these two can exist in downstream in packages such as RPMs. For example, the the RHEL system roles co-exit with the  `RHEL system roles collection <https://github.com/maxamillion/collection-rhel-system-roles>`_ and providing existing backwards compatibility with the downstream RPM.

Reviewer 4: Something to note here, that repo of mine is just an *example* of grabbing a single release of RHEL System Roles and "porting" it to Roles inside Collections and creating a spec file that supports both Standalone Roles and Roles in Collections, it is not in any way the authoritative location of RHEL System Roles. I want to make sure that's explicit here just to make sure there's no confusion on the topic.

  %description
  Ansible Collection and standalone role (for backward compatibility and migration) to deploy, configure, and manage the ACME webapp software.

  %prep
  %setup -qc

  %build

  %install

  mkdir -p %{buildroot}/%{collection_dir}
  cp -r ./* %{buildroot}/%{collection_dir}/

  mkdir -p %{buildroot}/%{_datadir}/ansible/roles
  for role in %{buildroot}/%{collection_dir}/roles/*
    do
	   cp -pR ${role} %{buildroot}/%{_datadir}/ansible/roles/%{roleprefix}$(basename ${role})

	   mkdir -p %{buildroot}/%{_pkgdocdir}/$(basename ${role})
	   for docfile in README.md COPYING LICENSE
	    do
      	if [ -f ${role}/${docfile} ]
    	    then
          	cp -p ${role}/${docfile} %{buildroot}/%{_pkgdocdir}/$(basename ${role})/${docfile}
      	fi
	   done
  done

  %files
  %dir %{_datadir}/ansible
  %dir %{_datadir}/ansible/roles
  %dir %{_datadir}/ansible/collections
  %dir %{_datadir}/ansible/collections/ansible_collections
  %{_datadir}/ansible/roles/
  %doc %{_pkgdocdir}/*/README.md
  %doc %{_datadir}/ansible/roles/%{roleprefix}*/README.md
  %{collection_dir}
  %doc %{collection_dir}/roles/*/README.md
  %license %{_pkgdocdir}/*/COPYING

>   %license %{_pkgdocdir}/*/LICENSE

Reviewer 2: https://github.com/ansible-community/collection_migration/blob/master/migrate.md has several missing steps, specially for 'non modules'. some points only apply to migrating from core, but others apply to any/all moves of a plugin

Author: I don't know what you are asking for in this comment?

Author: ok found it and moved some of them over. thanks!

See :ref:`developing_collections` for details on collections.

.. contents::
   :local:
   :depth: 1

Comparing standalone roles to collection roles
===============================================

:ref:`Standalone roles <playbooks_reuse_roles>` have the following directory structure:

.. code-block:: bash
  :emphasize-lines: 5,7,8

    role/
    ├── defaults
    ├── files
    ├── handlers
    ├── library
    ├── meta
    ├── module_utils
    ├── [*_plugins]
    ├── tasks
    ├── templates
    ├── tests
    └── vars

The highlighted directories above will change when you migrate to a collection-based role. The collection directory structure includes a :file:`roles/` directory:

.. code-block:: bash

    mynamespace/
    └── mycollection/
      ├── docs/
      ├── galaxy.yml
      ├── plugins/
      │   ├── modules/
      │   │   └── module1.py
      │   ├── inventory/
      │   └── .../
      ├── README.md
      ├── roles/
      │   ├── role1/
      │   ├── role2/
      │   └── .../
      ├── playbooks/
      │   ├── files/
      │   ├── vars/
      │   ├── templates/
      │   └── tasks/
      └── tests/

> You will need to use the Fully Qualified Collection Name (FQCN)` when you migrate your role into a collection. The FQCN is the combination of the collection ``namespace``, collection ``name``, and the content item you are referring to.

Reviewer 3: Is this line missing a ` somewhere?

> Fully Qualified Collection Name (FQCN)`

Reviewer: oh, yep, thanks

> You will need to use the Fully Qualified Collection Name (FQCN)` when you migrate your role into a collection. The FQCN is the combination of the collection ``namespace``, collection ``name``, and the content item you are referring to.

Reviewer: ```suggestion
You will need to use the Fully Qualified Collection Name (FQCN) when you migrate your role into a collection. The FQCN is the combination of the collection ``namespace``, collection ``name``, and the content item you are referring to.
```

Reviewer: ah, this was already resolved, closing

So for example, in the above collection, the FQCN to access  ``role1`` would be:

.. code-block:: Python

	mynamespace.mycollection.role1

A collection can contain one or more roles in the :file:`roles/` directory and these are almost identical to standalone roles, except you need to move plugins out of the individual roles, and use the :abbr:`FQCN (Fully Qualified Collection Name)` in some places, as detailed in the next section.

.. note::

	 In standalone roles, some of the plugin directories referenced their plugin types in the plural sense; this is not the case in collections.

.. _simple_roles_in_collections:

Migrating a role to a collection
=================================

To migrate from a standalone role (with no plugins) to a collection role:

1. Create a local :file:`ansible_collections` directory and ``cd`` to this new directory.

> 2. Create a collection. You need a `Galaxy namespace <https://galaxy.ansible.com/docs/contributing/namespaces.html>`_ to import this collection to Galaxy.

Reviewer 3: I know we talked about this previously—but as this guide seems to be for any role (and not just roles that are already on Galaxy)—should this be a top level requirement, or can we move it to an aside? If an aside, I'd suggest:

> ```
> If you wish to share your collection on Ansible Galaxy, you will need a `Galaxy namespace <https://galaxy.ansible.com/docs/contributing/namespaces.html>` to import the collection. Make sure you use the same namespace when creating the collection.
> ```

Something along those lines.

.. _migrating_roles:

*************************************************
Migrating Roles to Roles in Collections on Galaxy
*************************************************

> You can migrate an existing standalone role into a collection-based role that you can host on Galaxy. This offers you more flexibility and resources to have your role within a collection. With Ansible collections on Galaxy, you can use a single collection as a delivery mechanism for many roles, releasing them as a single cohesive unit of re-usable automation. Inside a collection, custom plugins that used to be shipped in the role's :file:`library/`` directory can now be shared across all roles within a

Reviewer: ```suggestion
You can migrate any existing standalone role into a collection and host the collection on Galaxy. With Ansible collections, you can distribute many roles in a single cohesive unit of re-usable automation. Inside a collection, you can share custom plugins across all roles in the collection instead of duplicating them in each role's :file:`library/`` directory.
```

> collection without duplicating them in each role's :file:`library/`` directory.

Reviewer: ```suggestion
```

> You also need to migrate roles to collections if you want them as part of certified Ansible content.

Reviewer: ```suggestion
You must migrate roles to collections if you want to distribute them as certified Ansible content.
```

.. note::

> 	If you want to import your collection to Galaxy, you need a `Galaxy namespace <https://galaxy.ansible.com/docs/contributing/namespaces.html>`_.

Reviewer: I might move this to a "Prerequisites" section right above the migration instructions.

.. code-block:: bash

    mynamespace/
    └── mycollection/
      ├── docs/
      ├── galaxy.yml
      ├── plugins/
      │   ├── modules/
      │   │   └── module1.py
      │   ├── inventory/
      │   └── .../
      ├── README.md
      ├── roles/
      │   ├── role1/
      │   ├── role2/
      │   └── .../
      ├── playbooks/
      │   ├── files/
      │   ├── vars/
      │   ├── templates/
      │   └── tasks/
      └── tests/

You will need to use the Fully Qualified Collection Name (FQCN) to use the roles and plugins when you migrate your role into a collection. The FQCN is the combination of the collection ``namespace``, collection ``name``, and the content item you are referring to.

So for example, in the above collection, the FQCN to access  ``role1`` would be:

.. code-block:: Python

	mynamespace.mycollection.role1

A collection can contain one or more roles in the :file:`roles/` directory and these are almost identical to standalone roles, except you need to move plugins out of the individual roles, and use the :abbr:`FQCN (Fully Qualified Collection Name)` in some places, as detailed in the next section.

.. note::

> 	 In standalone roles, some of the plugin directories referenced their plugin types in the plural sense; this is not the case in collections.

Reviewer: Is this true? In the example above, we have `modules/` in the `plugins/` directory - that's plural.

.. _simple_roles_in_collections:

Migrating a role to a collection
=================================

> To migrate from a standalone role (with no plugins) to a collection role:

Reviewer: ```suggestion
To migrate from a standalone role that contains no plugins to a collection role:
```

5. Move any other plugins to the appropriate :file:`plugins/PLUGINTYPE/` directory.  See :ref:`migrating_plugins_collection` for additional steps that may be required.

6. Update ``galaxy.yml`` to include any role dependencies.

7. Update the collection README.md file to add links to any role README.md files.

8. Change any references to the role to use the :abbr:`FQCN (Fully Qualified Collection Name)`.

.. code-block:: yaml

  ---
  - name: example role by FQCN
    hosts: some_host_pattern
    tasks:
      - name: import FQCN role from a collection
        import_role:
          name: mynamespace.mycollection.my_role

You can alternately use the ``collections`` keyword to simplify this:

.. code-block:: yaml

  ---
  - name: example role by FQCN
    hosts: some_host_pattern
    collections:
      - mynamespace.mycollection
    tasks:
      - name: import role from a collection
        import_role:
          name: my_role

.. _migrating_plugins_collection:

Migrating other role plugins to a collection
---------------------------------------------

To migrate other role plugins to a collection:

> 1. Move any nonmodule plugins to the appropriate :file:`plugins/PLUGINTYPE/` directory. The :file:`mynamespace/mycollection/plugins/README.md` file explains the types of plugins that the collection can contain within optionally created subdirectories.

Reviewer: ```suggestion
1. Move each nonmodule plugins to the appropriate :file:`plugins/PLUGINTYPE/` directory. The :file:`mynamespace/mycollection/plugins/README.md` file explains the types of plugins that the collection can contain within optionally created subdirectories.
```
